### ISSUE TITLE
Make `Witness::element_at` public

### ISSUE THREAD

**Issue by Tibo-lg** [rust-bitcoin/rust-bitcoin#1047]
I'm wondering if there is a reason for the [`Witness::element_at`](https://github.com/rust-bitcoin/rust-bitcoin/blob/master/src/blockdata/witness.rs#L227) method not to be public? Now I need to ~either reimplement it in my code or~ use `to_vec` first. Happy to PR the change but wanted to check first if maybe there was a reason for it being private at the moment.

**Comment by RCasatta** [comment]
There was a discussion on implementing `Index` (same effect as having `pub element_at`) on witness but it wasn't implemented to prevent inefficient (if you look at inner implementation accessing `ith` element requires iterating the inner content) access like:

```
for i in 0..witness.len() {
   print("{:?}",witness[i])
}
```

an alternative to use `witness.to_vec(i)` is `witness.iter().nth(i)` which better convey the inefficiency. Note also it's common to access last or second-to-last element in the witness and there are efficient methods for those `last()` and `second_to_last()`

**Comment by Tibo-lg** [comment]
Thanks for the reply and explanation, I had not been able to find any discussion related to that through the github search. Unfortunately `last()` and `second_to_last()` don't work for me. Basically the reason I need to access the witness is to retrieve a signature from a multi sig script, which can be in 2nd or 3rd position (out of four witness elements). I guess I'll live with it and it's not like it's gonna impact performance of my application, but it just feels a bit awkward.

**Comment by Kixunil** [comment]
I don't see how leaking internal representation would help here. There's no way to avoid linear search or conversion by doing that.

**Comment by apoelstra** [comment]
@Kixunil if you know your witness stack has a particular form (e.g. a series of 65-byte signatures) you can exploit random access.

**Comment by Kixunil** [comment]
You'd have to check it first by iterating all elements if the data came from external source at which point you may as well convert it into custom type. If it's internal you can just avoid this type (or convert when serializing).

But thinking about it, maybe we could have consuming conversions from/to `Vec<u8>`. If we decide to change the internal representation we can keep them, although they could become slower.

**Comment by apoelstra** [comment]
Even if the witness came from an external source (not guaranteed), if you knew (a) the witness program required a specific form of witness, and (b) that the witness is valid (e.g. it came from the blockchain), you can still be assured of the shape of the witness.

**Comment by Kixunil** [comment]
Ah, good point. Still, I think conversions are best for these things. Or if people have some interesting use cases I wouldn't mind trying to figure out a better API.

**Comment by apoelstra** [comment]
I agree conversions are better in principle, but in this case we could return a borrowed slice to the inner data, which (a) would be extremely cheap, and (b) will inevitably expose the inner representation, no matter how much we don't want to.

**Comment by sanket1729** [comment]
For the record, I need this to test that my witness is actually of the form that I expect it to be. I am currently using `to_vec`.

> I guess I'll live with it and it's not like it's gonna impact performance of my application, but it just feels a bit awkward.

It feels awkward, but is not that bad. 
1) If you parsing a witness from a blockchain, doing a `to_vec` is not bad compared to the previous implementation of parsing a `Vec<Vec<u8>>` directly. See [this](https://github.com/rust-bitcoin/rust-bitcoin/pull/672) for benchmarks
 
2) If you are creating a witness, you can create a Vec<Vec> initially and later convert using `from_vec`

**Comment by Tibo-lg** [comment]
Again it's not a big deal for me, but just to add to the discussion:

@apoelstra the cleanest way to expose something useful and easy to use IMHO would be to implement `Index` or have something like a public `element_at` method. Both would be cheap and avoid exposing internal I think. I also don't think preventing the users from doing something inefficient by forcing them to do something inefficient in other cases is a good argument, but 🤷 .

@sanket1729 I agree it's not that bad, it's just not really intuitive not immediately obvious that you should use `to_vec` or `iter().nth()` to retrieve an element.

If anybody is interested, I made some experiments to see how different approaches perform: https://github.com/p2pderivatives/rust-bitcoin/tree/witness-experiments
(I probably have better things to do, but got carried away)

**Comment by Kixunil** [comment]
@Tibo-lg `Index` is really no-go for the same reason it's not implemented on `String`. I think what we really need to do is to improve the documentation. If you need your algorithm to be efficient we may be able to come up with a better API when we understand what exactly you're doing.

**Comment by Tibo-lg** [comment]
@Kixunil Note I'm not trying to push anything (I closed the issue), just giving my opinion reading the on-going discussion.

> for the same reason it's not implemented on String.

I get the similarity with `String` in that it's not trivial to get an element by index, but compared to a `String`, it feels like getting an element from the witness stack is one of the main thing one may want to do with a `Witness` struct.

> I think what we really need to do is to improve the documentation.

I think that's a very good point. Maybe mentioning `.iter().nth()` somewhere would be useful (my first thought when I realized that there was no easy way was to use `to_vec`). 

> If you need your algorithm to be efficient we may be able to come up with a better API when we understand what exactly you're doing.

I don't, I mentioned what I'm doing [here](https://github.com/rust-bitcoin/rust-bitcoin/issues/1047#issuecomment-1150751882), it's really not more about ease of use than efficiency.

**Comment by Kixunil** [comment]
If someone wants to try adding `indices: Vec<usize>` field to `Witness`, implement `Index` as `element_at(indices[index])` and bench how much it changes I'd be happy to review such a PR and accept if benches are interesting.

Also #1050 is related.

**Comment by Tibo-lg** [comment]
> If someone wants to try adding `indices: Vec<usize>` field to `Witness`, implement `Index` as `element_at(indices[index])` and bench how much it changes I'd be happy to review such a PR and accept if benches are interesting.
> 
> Also #1050 is related.

@Kixunil Gave a quick try to having an `indices` field but it does have a significant impact on deserialization (~15%) so probably not worth it. An alternative would be to have some lazy construction of `indices` (something like if you request at index 3 it stores the indices for 1 and 2 since it has to go through them anyway).

**Comment by Kixunil** [comment]
@Tibo-lg Thanks for looking at it! Just to be sure, did you preallocate the indices `Vec`?

**Comment by Tibo-lg** [comment]
> @Tibo-lg Thanks for looking at it! Just to be sure, did you preallocate the indices `Vec`?

Yes, I tried with both creating the `Vec` with `with_capacity` and the `vec!` macro but it doesn't seem to make a big difference. Variance is also quite big so it's difficult to say exactly, but running quite several times I consistently get the average at least +10% on the version with `indices`.

I've pushed the implementation [here](https://github.com/p2pderivatives/rust-bitcoin/tree/witness-experiments-2) if you want to have a look.
